### DISCUSSION TITLE
How to use QUIC protocol for EMQX 5.0 in K8S network

### DISCUSSION THREAD

**Discussion by Rory-Z** [emqx/emqx-operator#490] [Q&A] [answered]
EMQX 5.0 provides `MQTT over QUIC` feature, and I want use this in K8S network by EMQX Operator, what should I do ?

**Comment by Gala-R** [comment, marked as answer]
# QUIC for EMQX 5.0 on K8S 

## Environment

|     EMQX Operator |  Kubernetes  |        Kernel               |              OS                |
|:----------------:|:------------:|:---------------------------:|:------------------------------:|
|       2.0.2      |    1.25.2    |  5.4.0-51-generic           | Ubuntu Server 20.04 LTS 64-bit |

**NOTE**: When kube-proxy is set to ipvs mode and UDP service is exposed in NodePort mode, the UDP service will be disconnected. This is a known bug. For specific issues, please refer to: [kube-proxy IPVS break UDP NodePort Services by clearing active conntrack entries](https://github.com/kubernetes/kubernetes/issues/113802). This bug has been fixed in K8s 1.27, please refer to: [Syncing IPVS conntrack cleaning with IPTables](https://github.com/kubernetes/kubernetes/pull/116171).

## EMQX configuration

MQTT over QUIC is an experimental function of EMQX 5.0. For its documentation, please refer to: [MQTT Over QUIC](https://www.emqx.io/docs/zh/v5.0/advanced/quic.html#%E7%AE%80%E4%BB%8B).

EMQX CRD supports the use of `.spec.bootstrapConfig` to configure the EMQX cluster. This configuration will be used as the EMQX bootstrap config, and its format is consistent with the emqx.conf file, both in hocon format. This field is only allowed to be configured when creating an EMQX cluster, and does not support updating. The QUIC listener is disabled by default, so if you want to use QUIC, you need to enable QUIC in `.spec.bootstrapConfig` configuration.

```yaml
apiVersion: apps.emqx.io/v2alpha1
kind: EMQX
metadata:
   name: emqx
spec:
   image: emqx/emqx:5.0.9
   bootstrapConfig: |
     listeners.quic.default {
      enabled = true
      bind = "0.0.0.0:14567"
      keyfile = "/opt/emqx/etc/certs/key.pem"
      certfile = "/opt/emqx/etc/certs/cert.pem"
     }
   coreTemplate:
     spec:
       replicas: 3
   replicantTemplate:
     spec:
       replicas: 3
   listenersServiceTemplate:
     spec:
       type: NodePort
       ports:
         - name: quic-default
           protocol: UDP
           port: 14567
           targetPort: 14567
```

**NOTE**: `enabled` is set to true to enable the QUIC listener. `bind` indicates the IP and port that the QUIC listener listens on. `keyfile` indicates the server private key file path. `certfile` indicates the server certificate file path.

Save the above file as: emqx-quic.yaml, and execute the following command to deploy the EMQX cluster

```
kubectl apply -f emqx-quic.yaml
````

The output is similar to:

```
emqx.apps.emqx.io/emqx created
```

- Wait for EMQX cluster to be ready, you can check the status of EMQX cluster through the `kubectl get` command, please make sure that `STATUS` is `Running`, this may take some time

```
$ kubectl get emqx
```

The output is similar to:

```
NAME   IMAGE      STATUS    AGE
emqx   emqx:5.0   Running   10m
```

**NOTE**: `node` represents the unique identifier of the EMQX node in the cluster. `node_status` indicates the status of the EMQX node. `otp_release` indicates the version of Erlang used by EMQX. `role` represents the EMQX node role type. `version` indicates the EMQX version. EMQX Operator creates an EMQX cluster with three core nodes and three replicant nodes by default, so when the cluster is running normally, you can see information about three running core nodes and three replicant nodes. If you configure the `.spec.coreTemplate.spec.replicas` field, when the cluster is running normally, the number of running core nodes displayed in the output should be equal to the value of this replicas. If you configure the `.spec.replicantTemplate.spec.replicas` field, when the cluster is running normally, the number of running replicant nodes displayed in the output should be equal to the replicas value.

- Check if EMQX QUIC listener is ready

```
kubectl get svc -l apps.emqx.io/instance=emqx
```

The output is similar to:

```
NAME             TYPE        CLUSTER-IP       EXTERNAL-IP   PORT(S)                          AGE
emqx-dashboard   ClusterIP   10.105.67.95     <none>        18083/TCP                        18m
emqx-listeners   NodePort    10.102.204.194   <none>        14567:32250/UDP,1883:32402/TCP   17m
```

**NOTE**: `Name` means the name of the Service, and EMQX Operator will create a Servcie named emqx-dashboard by default to bind port 18083 of the EMQX Dashboard service. The Service named emqx-listeners will decide whether to create it according to whether the user has configured `.spec.listenersServiceTemplate` and whether EMQX has enabled the listener. EMQX Operator will automatically merge the user-configured listener and EMQX enabled according to the name of the listener of the listener. `TYPE` indicates the type of Service. In this article, the type of emqx-listeners Serivce configuration is NodePort. For the documentation on the type of Service, please refer to: [Service Type](https://kubernetes.io/zh-cn/docs/concepts/services-networking/service/#publishing-services-service-types). `PORT(S)` indicates the listener port and protocol, and the corresponding nodePort port.

## Test QUIC based on emqtt-bench

[emqtt-bench](https://github.com/emqx/emqtt-bench) is written based on Erlang, a simple and powerful MQTT protocol performance testing tool. Users can go to [emqtt-bench releases](https://github.com/emqx/emqtt-bench/releases) to download and install emqtt-bench that supports the QUIC protocol.

- Subscribe message test

```
./emqtt_bench sub --quic true -h 192.168.1.208 -p 32250 -c 10 -t "test"
```

**NOTE**: `./emqtt_bench` indicates the installation path of emqtt-bench. `sub` means a subscription request. `--quic true` means to enable the QUIC protocol. `-h` indicates the destination IP address of the request. In this article, the way the EMQX cluster exposes services to the outside world is NodePort, so the IP address here is the IP of any node in the Kubernetes cluster. `-p` indicates the destination port, which is 32250 in this article. `-c` indicates the number of clients. This article configures 10 clients to subscribe to messages. `-t` indicates the subscribed topic, the subscribed topic in this article is test. For more detailed use of emqtt-bench, please refer to the document: [emqtt-bench](https://github.com/emqx/emqtt-bench/blob/master/README.md)

- Publish message test

```
./emqtt_bench pub --quic true -h 192.168.1.208 -p 32250 -c 1 -t "test" -m "hello world"
```

**NOTE**: The command of emqtt-bench to publish a message is very similar to the command to subscribe to a message, `pub` means a request to publish a message. `-m` indicates the message body.
